### PR TITLE
RFC: Drop dimensions indexed by scalars

### PR DESCRIPTION
**Pull request by mbauman**
This is the first step towards APL indexing semantics.  It does not yet allow indexing with multiple multi-dimensional arrays (the new feature with the catchy "rank of the result is the sum of the rank of the indices" semantic); it simply drops all scalar dimensions (a major breaking change).

There is an interesting interaction between these array semantics and transposes that I didn't fully appreciate until looking at this more in-depth.  The tricky part about APL indexing is that it loses the orientation of row vectors, and so we can no longer define complex matrix multiplication in terms of slices directly:
- This makes `A[1,:] * B[:,1]` an illegal `(n,) × (n,)` operation.
- Similarly, `dot(A[1,:], B[:,1])` and `A[1,:]' * B[:,1]` erroneously conjugate the row.

Working with complex row slices becomes a little dicier. When do you conjugate? Or (c)transpose?

Fortunately, the behavior here is a strict superset of our current indexing rules.  You can explicitly request that the shapes of the slices be maintained in both dimensions: `A[1:1,:] * B[:,1:1]` if you wish to use the result as a linear algebra construct.

### CODE REVIEW

**Comment by mbauman** [comment]
AppVeyor failure looks like #9572.

File: base/linalg/diagonal.jl [hunk 1]
```diff
@@ -121,7 +121,10 @@ function A_ldiv_B!{T}(D::Diagonal{T}, V::AbstractMatrix{T})
         if d == zero(T)
             throw(SingularException(i))
         end
-        V[i,:] *= inv(d)
+        d⁻¹ = inv(d)
```
Comment (IainNZ) [inline base/linalg/diagonal.jl]: Very cute

Comment (malmaud) [inline base/linalg/diagonal.jl]: :koala:

**Comment by andreasnoack** [comment]
@mbauman Would you rebase this one. I think we should consider merging this soon such that people will start using it.

**Comment by mbauman** [comment]
Done.

**Comment by mschauer** [comment]
Thanks for keeping it rolling here

**Comment by andreasnoack** [comment]
Okay to merge this?

**Comment by mbauman** [comment]
Sure, let's give it a shot.

**Comment by timholy** [comment]
Sorry I've not had time to review this. Looks fine to me. As you noted, it depends on a lot more splatting (esp in checksize), but all that is O(1) and not O(N) in the array elements. So I'm fine with seeing how it goes (and it might be further incentive to fix the splatting penalty), though if you've run any benchmarks it might be interesting to learn more about any performance hits.

**Comment by yuyichao** [comment]
Should we also drop the scalar dimension for `SubArray` (especially if we'd like to return `SubArray` as the result of indexing).

This breaks the equivalence between `getindex` and `sub`

```jl
julia> a = rand(2, 2)
2x2 Array{Float64,2}:
 0.377862   0.85667 
 0.0177678  0.695366

julia> a[2, 1:2] == sub(a, 2, 1:2)
false
```

**Comment by timholy** [comment]
`slice` already does that---essentially, now `getindex` is equivalent to `slice` rather than `sub`.

**Comment by mschauer** [comment]
Are there plans to adapt ``mapslices`` to ``slice`` rather than to ``sub``?

**Comment by andreasnoack** [comment]
I don't think it has been discussed. Can you explain further?

**Comment by mschauer** [comment]
Currently, ``mapslices`` and special cased reductions over dimensions keep the reduced dimension as singleton dimension. As the indexing behavior now changed I thought mapslices and friends should follow soonish. This would mean changing the standard behavior of mapslices to dimension dropping as hinted at by https://github.com/JuliaLang/julia/blob/dc5c974df77129f3d129014e0acdb1997aeade8b/base/abstractarray.jl#L1217

**Comment by timholy** [comment]
I'm not sure I agree that those two design decisions are coupled. Meaning, I don't think it's inconsistent for `getindex` to drop dimensions but reductions to retain them: they are different function calls, and reductions are obviously different from indexing.

It's so common to say
```jl
Xnorm = X ./ maximum(X, 2)
```
that it would be a shame to make that any harder. (And dropping dimensions definitely does make that harder.) When you want to drop the dimensions, it's easier than ever to do so.

**Comment by StefanKarpinski** [comment]
I agree – I would prefer that reductions keep reduced dimensions for exactly this reason.

**Comment by mschauer** [comment]
I kind of agree for the monadic case, for a dyadic mapslices function the current scheme does not translate well. I now recall why I noted this, I have even RFC #13317 but I forgot about it. Maybe someone could comment there? Some feedback would be appreciated.

**Comment by stevengj** [comment]
I think this should be moved to the "Breaking Changes" part of the NEWS file.  This just caused silently incorrect results in one of my packages (Hadamard.jl)

**Comment by davidavdav** [comment]
Hello, 

For `NamedArrays` I need to copy the exact behaviour of the dropping of dimensions (for julia-0.4 and 0.5).  The AbstractArray contained in a `n::NamedArray` is sliced by calling `getindex(n.array, index...)`---so there I get the same behaviour automatically.  But for the names of the dimensions in `n`, I need to know exactly which types of `index...` I need to keep.

Is there a `Base` function that will tell me this?  I was looking at `Base.index_shape` but that isn't defined for all index types (e.g., `CartesianIndex`).  Currently I use a helper function
```julia
dimkeepingtype(x) = false
dimkeepingtype(x::Vector) = true
dimkeepingtype(x::Range) = true
dimkeepingtype(x::BitVector) = true
```
that tells me if an index type `x ` (∈ `index`) is non-scalar or not.  I am not sure this covers all index types in the same way that `AbstractArray` keeps dimensions, so I'd rather refer similar functionality in `Base`.

**Comment by timholy** [comment]
Can't you add the missing `index_shape_dim` method(s) to NamedArrays? For julia-0.5 we could add such a method to Base, though---probably a good idea.

**Comment by davidavdav** [comment]
Thanks.  But I do not really understand the logic in `index_shape_dim`---it seems a recursive helper function for `index_shape`.  It doesn't look like the way towards telling me which dimensions are kept in slices.  I need these to know which dimension names I need to select, and from which dimensions I have to slice the names of the indices in the `NamedArray`.   

Somewhat related, I saw that in Julia-0.5 the dimension of a sliced array is the sum of the dimensions of the indices.  That is great, but I don't really know yet how to come up with dimension and index names, with cases like `NamedArray[Array]` and `NamedArray[NamedArray]`.   Does anybody know of a precedence for this case?  I only know of R named arrays, but R doesn't have this fancy indexing scheme.

**Comment by mbauman** [comment]
I don't know about precedence in other languages, but perhaps you can take some inspiration from AxisArrays:

* `AxisArray[Array1, Array1]` returns an AxisArray with axis names `row_1`, `row_2`, …, `row_$(ndims(Array1))`, and `col_1`, `col_2`, …, `col_$(ndims(Array2))`.
* `AxisArray[AxisArray]` is similar, except instead of using numbers on the right-hand side of the `_`, it uses the axis names of the index.  This is where the axis names `time_sub` and `time_rep` come from on the README example.

**Comment by davidavdav** [comment]
Thanks, that is indeed a useful way of naming the dimensions. 

For the names of the indexes along the dimensions, though, things are more complicated.  I think I will opt for copying the names of the indexes of the indexing array (if this has `ndims() > 1`), or defaulting to 1..size(index) if the indexing array itself doesn't have names. 

Further challenges will occur when indexing a `NamedArray` with an `AxisArray` or vice versa.  I don't even want to think about this at this moment...

**Comment by floswald** [comment]
i'm not sure whether this is intended but I think `findmax` does not conform to this:

```julia
julia> x = rand(2,2,2)
2×2×2 Array{Float64,3}:
[:, :, 1] =
 0.319551  0.616412
 0.740493  0.112079

[:, :, 2] =
 0.0639773  0.376866
 0.475386   0.46051 

julia> z = findmax(x,2);

julia> z[1]
2×1×2 Array{Float64,3}:
[:, :, 1] =
 0.616412
 0.740493

[:, :, 2] =
 0.376866
 0.475386
```
I would expect a `(2,2)`, i.e. the scalar dim should be `squeeze`d out.

**Pull request merged**